### ISSUE TITLE
Trouble with k-fold CV in MSOMs

### ISSUE THREAD

**Issue by zacksteel** [doserjef/spOccupancy#34]
Thanks for the great package. I'm trying to run k-fold cross validation with a MSOM but am receiving some errors. 

With my own dataset and model I get this at the beginning of the CV process:
```
Performing 5-fold cross-validation using 5 thread(s).
Error in { : task 1 failed - "(subscript) logical subscript too long"
```

I tried running the examples in the msPGOcc help documentation and the MSOM example in the [intro vignette](https://www.jeffdoser.com/files/spoccupancy-web/articles/modelfitting#model-selection-using-waic-and-k-fold-cross-validation-1). In both cases I get this error at the end of the run (when compiling the returned object?):
```
Performing 5-fold cross-validation using 1 thread(s).
Error in { : task 1 failed - "object 'coef.names.det' not found"
```

The single-species model in the vignette runs fine w/ k-fold CV. The two errors above seem to be different issues, but I thought I'd report both here. With my dataset and the full model runs fine. I can provide the code and/or the data if helpful for diagnosing the CV error.

Thanks again,
Zack

**Comment by doserjef** [comment]
Hi @zacksteel,

Thanks very much for the note, and sorry you're running into these issues. I just ran the `msPGOcc()` example in the documentation with k-fold CV as well as the example in the intro vignette with the Hubbard Brook data and they both ran without any errors. There were some issues in previous versions of the package with the k-fold CV, so perhaps the simplest place to start would be to re-install the package from CRAN to make sure that you're working with the latest version of the package (v0.7.2), which is what I just tested it with. 

If reinstalling the latest version doesn't help, then please email me your data and code (doserjef@msu.edu) and I'll dig into it further to try and see what's going on. 

Thanks, and happy new year!

Jeff

**Comment by zacksteel** [comment]
Thanks for the response @doserjef. Unfortunately, it does not seem to be a version issue. I've tried running the vignette and HB examples with spOccupancy v0.7.2 on three different machines (two windows OSs and one Mac with a mix of the latest version of R and older versions). In all cases I still get the _task 1 failed - "object 'coef.names.det' not found"_ error.

For my data and model I think I've isolated what is causing the _"(subscript) logical subscript too long"_ error. I have a predictor in the detection model that results in an error when included in the model, but when excluded the error does not occur. This predictor has some missing values, which I suspect is causing the problem. When I fill in these blanks the error goes away. Strangely, msPGOcc allows missing values in the detection submodel when not running CV. In contrast, when introducing missing values in a predictor on occupancy, msPGOcc will throw an error (as I expected it would). As with the vignette examples, I still get the coef.names.det error at what I believe is the end of the CV run.

Happy New year!
Zack

**Comment by doserjef** [comment]
Ah bummer that it wasn't just a version issue. Thanks for testing on multiple machines. I haven't been able to reproduce the error yet with the vignette/HB examples, but I'll keep testing and get back to you if I figure out what's going on. 

Glad to hear you isolated the problem with your data set, although I agree that is some weird behavior with the detection predictor. You're correct that msPGOcc should throw an error if there are missing values in any occupancy covariates. The function should be able to handle missing values in the detection covariates (both for fitting the model and doing CV), so long as they line up in a predictable way with any missing values in the detection-nondetection data values. If possible, could you send me your data/code for me to play around with so I can try to reproduce this and fix it? 

Thanks!

Jeff

**Comment by zacksteel** [comment]
Ok, I think I see what we are doing differently (maybe). I've been running these models with the k.fold.only = T when getting the "object 'coef.names.det' not found" error. When I set that parameter to F everything runs fine. Is the output model object not fully set up if only running the k fold step and not the full model?

I'm sending my code and data to your email if you'd like to diagnose the other error.

Thanks again,
Zack

p.s. there appears to be a typo in your documentation of the k.fold.only parameter in the single-species occupancy models section of the vignette. It currently reads: _Lastly, the k.fold.only argument controls whether the full model is run prior to performing cross-validation (k.fold.only = FALSE, the default). If set to FALSE, only k-fold cross-validation will be performed._ I'm assuming the second sentence is intended to read "If set to TRUE..."

**Comment by doserjef** [comment]
Ah okay, thanks a lot for the clarification, you're correct that I wasn't running things with `k.fold.only = TRUE`. I was able to reproduce the error when `k.fold.only = TRUE`, which occurred in all the multi-species occupancy model functions. I fixed the problem in the development version on GitHub, so if you re-install from GitHub with `devtools::install_github("doserjef/spOccupancy")` then cross-validation should now work for multi-species models when `k.fold.only = TRUE`. Let me know if that doesn't work for you.

Thanks again for pointing this out! I'm going to leave this issue open until I get the updated version up on CRAN. 

Jeff
